linode/provider.go

```diff
@@ -251,31 +233,90 @@ func Provider() *schema.Provider {
 	return provider
 }
 
+func handleDefault(config *helper.Config, d *schema.ResourceData) diag.Diagnostics {
+	if v, ok := d.GetOk("token"); ok {
+		config.AccessToken = v.(string)
+	} else {
+		config.AccessToken = os.Getenv("LINODE_TOKEN")
+	}
+
+	if v, ok := d.GetOk("api_version"); ok {
+		config.APIVersion = v.(string)
+	} else {
+		config.APIVersion = os.Getenv("LINODE_API_VERSION")
+	}
+
+	if v, ok := d.GetOk("config_path"); ok {
+		config.ConfigPath = v.(string)
+	} else {
+		homeDir, err := os.UserHomeDir()
+		if err != nil {
+			return diag.Errorf(
+				"Failed to get user home directory: %s",
+				err.Error(),
+			)
+		}
+		config.ConfigPath = fmt.Sprintf("%s/.config/linode", homeDir)
+	}
+
+	if v, ok := d.GetOk("config_profile"); ok {
+		config.ConfigProfile = v.(string)
+	} else {
+		config.ConfigProfile = "default"
+	}
+
+	if v, ok := d.GetOk("url"); ok {
+		config.APIURL = v.(string)
+	} else {
+		config.APIURL = os.Getenv("LINODE_URL")
+	}
+
+	if v, ok := d.GetOk("ua_prefix"); ok {
+		config.UAPrefix = v.(string)
+	} else {
+		config.UAPrefix = os.Getenv("LINODE_UA_PREFIX")
+	}
+
+	if v, ok := d.GetOk("event_poll_ms"); ok {
+		config.EventPollMilliseconds = v.(int)
+	} else {
+		eventPollMs, err := strconv.ParseInt(os.Getenv("LINODE_EVENT_POLL_MS"), 10, 64)
+		if err != nil {
+			eventPollMs = 4000
+		}
+		config.EventPollMilliseconds = int(eventPollMs)
+	}
+
+	if v, ok := d.GetOk("lke_event_poll_ms"); ok {
+		config.LKEEventPollMilliseconds = v.(int)
+	} else {
+		config.LKEEventPollMilliseconds = 3000
+	}
+
+	if v, ok := d.GetOk("lke_node_ready_poll_ms"); ok {
+		config.LKENodeReadyPollMilliseconds = v.(int)
+	} else {
+		config.LKENodeReadyPollMilliseconds = 3000
+	}
+
+	return nil
+}
+
 func providerConfigure(
 	ctx context.Context, d *schema.ResourceData, terraformVersion string,
 ) (interface{}, diag.Diagnostics) {
 	config := &helper.Config{
-		AccessToken: d.Get("token").(string),
-		APIURL:      d.Get("url").(string),
-		APIVersion:  d.Get("api_version").(string),
-		UAPrefix:    d.Get("ua_prefix").(string),
-
-		ConfigPath:    d.Get("config_path").(string),
-		ConfigProfile: d.Get("config_profile").(string),
-
 		SkipInstanceReadyPoll:  d.Get("skip_instance_ready_poll").(bool),
 		SkipInstanceDeletePoll: d.Get("skip_instance_delete_poll").(bool),
 
 		DisableInternalCache: d.Get("disable_internal_cache").(bool),
 
 		MinRetryDelayMilliseconds: d.Get("min_retry_delay_ms").(int),
 		MaxRetryDelayMilliseconds: d.Get("max_retry_delay_ms").(int),
```

Comment:
do we need to move these to `handleDefault` as well?

Reply:
`d.Get()` would return the default values of that golang type if the key exist in schema but not set by the user.
https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema#ResourceData.Get

The default value of`SkipInstanceReadyPoll`, `SkipInstanceDeletePoll`, `DisableInternalCache` we set in the schema were `false`, which is also the default value of `bool` in go lang. Which means when you define a `bool` variable without assigning any value to it in go lang, it would be a `false`.

So, if we don't set a default value of `false` to them on Terraform application layer, the `d.Get()` will still return the desired default value when the user don't configure this attribute. So, having a default value for them is not necessary.